### PR TITLE
Catalog of Things Console Application

### PR DESCRIPTION
**Pull request by MrBrN197**
## Video Demo 
https://drive.google.com/file/d/1c0y9PTj3eC5tfgFGw3dY9F4v1T8N5JEi/view?usp=sharing

 - ✔ Created a main.rb file that serves as console app entry-point.
 - ✔ Implement startup actions:
    - ☑ Present the user with a list of options to perform.
    - ☑ Let users choose an option.
    - ☑ If needed, ask for parameters for the option.
    - ☑ Implement a way to quit the app

- ✔ Created a `Game` class
- ✔ `Author` class with an association to the `Item` class.
    - ☑ `add_item` method in the `Author` class
    - ☑ `can_be_archived?` in the `Game` class
- ✔ Create `GameFunctions` module
    - ☑ "List of games"
    - ☑ "List all authors"
    - ☑ "Add a game"
- ✔ All data is preserved by saving collections in .json files.
- ✔ Created a schema.sql
   - ☑ Added `games` table
   - ☑ Added `authors` table
- ✔ Add unit tests.
  - ☑ Add test for `Game` class
  - ☑ Add test for `Author` class

- Unit tests for the MusicAlbum class
- Unit tests for the Genre class
- Unit tests for the Item class
- Book class
- Label class
- Book functions module with options to add books, list books, and labels, as well as save books to a JSON file
- Unit tests
- Schema file to create book and label tables

- Created Item class in a separate .rb file.
- Added all methods visible in the diagram provided in the guidelines;
Implemented methods that could  do the following:
- can_be_archived?() in the Item class and should return true if published_date is older than 10 years
or otherwise, it should return false and move_to_archive() in the Item class
- should reuse can_be_archived?() method
- should change the archived property to true if the result of the can_be_archived?() method is true
- should do nothing if the result of the can_be_archived?() method

- Created a musical album class with relevant properties
- Ensuring that all Genre class properties visible in the diagram should be defined and set up in the constructor method.
Implemented methods by:
- adding_item method in the Genre class
 - taking an instance of the Item class as an input
-  added the input item to the collection of items
- should added self as a property of the item object (by using the correct setter from the item object)
- could be_be_archived?() in the MusicAlbum class
- should override the method from the parent class
- should return true if parent's method returns true AND if on_spotify equals true
otherwise, it should return false
- Added unit tests for all implemented methods.
The following options were made  available:
- Listing all music albums
- Listing all genres (e.g 'Comedy', 'Thriller')
- Adding a music album
 - Ensured that all data should be preserved by saving collections in .json files.
- Created a schema.sql file with tables that will be analogical to the structure of the classes that you created:
- music_albums table (add all properties and associations from the parent Item class as table columns)
genres table

## Video Demo 
https://drive.google.com/file/d/1c0y9PTj3eC5tfgFGw3dY9F4v1T8N5JEi/view?usp=sharing

### CODE REVIEW

File: classes/author.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  def initialize(first_name, last_name)
+    @id = rand(1..100_000)
+    @items = []
+    @first_name = first_name.strip
+    @last_name = last_name.strip
+  end
+
+  def add_item(item)
+    item.add_author(self)
+  end
```
Comment (mikethreels) [inline classes/author.rb]: - [x] to set up the 1-to-many relationships for `Author` you need to push `item` to `@items` as well as you did in the `Genre` class as well.

File: classes/label.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  def initialize(title, color)
+    @id = rand(1000)
+    @title = title
+    @color = color
+    @items = []
+  end
+
+  def add_item(item)
+    item.add_label(self)
+  end
```
Comment (mikethreels) [inline classes/label.rb]: - [x] to set up the 1-to-many relationships for `Label` you need to push `item` to `@items` as well as you did in the `Genre` class as well.

File: classes/musical_album.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  def to_json(json)
+    json.generate({ on_spotify: @on_spotify, publish_date: @publish_date.year })
+  end
+
+  private
+
+  def can_be_archived?
+    super && @on_spotify == true
+  end
```
Comment (mikethreels) [inline classes/musical_album.rb:21]: - [ ] according to the requirements `can_be_archived` should return true if the parent's method  returns `true` OR if `@on_spotify` equals true, in this case, it's only returning `true` if both cases are true

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Team, 

Highlights:
- it's great to see you all in the presentation with your camera on and the audio is good as well👍 

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._
- [x] as already mentioned above the video is great, however, unfortunately, according to the requirements every student should mention a particular part of code which they are proud of, therefore I have to ask you to please record a new video presentation including this as well.
- [x] you included the "list all Genres, labels and authors option", which is great however when creating a book, game or music album, you are not given the option to add a genre label or author which makes the application not work as expected.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classes/musical_album.rb (hunk 3 above)
Comment (mikethreels) [reply classes/musical_album.rb:21]: Hi team, 
My sincere apologies but please ignore this comment I made a small mistake and mixed up the musical_album with another class

**Review by D4nielJ — APPROVED** [review]
## Hello @MrBrN197, @leolpaz and @Busiwa24, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear or you can always find me on slack!

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**